src/compiler/parser.ts

```diff
@@ -6327,7 +6328,16 @@ namespace Parser {
         Debug.assert(languageVariant !== LanguageVariant.JSX, "Type assertions should never be parsed in JSX; they should be parsed as comparisons or JSX elements/fragments.");
         const pos = getNodePos();
         parseExpected(SyntaxKind.LessThanToken);
+        const posLessThan = getNodePos();
         const type = parseType();
+
+        // <Type />: specialized diagnostic for a JSX element in a .ts file - and skip some excess reporting
+        if (token() === SyntaxKind.SlashToken && scanner.scan() === SyntaxKind.GreaterThanToken) {
+            nextToken();
+            parseErrorAt(posLessThan - 1, getNodePos(), Diagnostics.JSX_tags_are_not_permitted_in_ts_files_Did_you_mean_for_the_file_extension_to_be_tsx, tokenToString(SyntaxKind.GreaterThanToken));
+            return finishNode(factory.createTypeAssertion(type, createMissingNode(SyntaxKind.Identifier, /*reportAtCurrentPosition*/ false)), pos);
+        }
+
         parseExpected(SyntaxKind.GreaterThanToken);
         const expression = parseSimpleUnaryExpression();

```

Comment:
Right after this, you'd want to check `if (token() === SyntaxKind.Identifier)` and `tryParse` some function that combines `parseJSXChildren` and `parseJSXClosingElement`. If that succeeds, you'll need to rebuild the previously parsed nodes above into an opening element (somehow??), issue your error, and return a JSX element. *In theory* the function you `tryParse` could be extracted from the first `if` body of `parseJsxElementOrSelfClosingElementOrFragment`, although it may do too much error recovery to be usable in a possibly-erroneous position.

Aside: Edge's AI-suggestion features are getting more aggressive but boy howdy was it confused by that paragraph.

Anyway, `if (token() === SyntaxKind.LessThanToken)` here, you'd instead `tryParse(parseJSXClosingElement)`, which covers the `<x>y</x>` case.

I'm not yet sure how to reparse `<x>y` as JSX opening/child token. Maybe there's somewhere else in the parser that does that, or maybe you can create some error fields for TypeAssertion and just cram the body/close token there. Like I said, TypeAssertions are rarely used at this point, so making its structure weird or inefficient is not a huge problem.